target/ppc/translate/vsx-impl.c.inc

```diff
@@ -1258,6 +1211,54 @@ static bool trans_XXPERMR(DisasContext *ctx, arg_XX3 *a)
     return true;
 }
 
+static bool trans_XXPERMDI(DisasContext *ctx, arg_XX3_dm *a)
+{
+    TCGv_i64 t0, t1;
+
+    REQUIRE_INSNS_FLAGS2(ctx, VSX);
+    REQUIRE_VSX(ctx);
+
+    t0 = tcg_temp_new_i64();
+
+    if (unlikely(a->xt == a->xa || a->xt == a->xb)) {
```

Comment:
I know this was in the original code, but I wonder if it really is worth special casing this. What if we simply do what's in the ``unlikely`` path every time?

Reply:
I did a simple benchmark and the version with the special case was consistently faster, but not by much. I'll see if I get the numbers again to post here.

Reply:
That's interesting. The only difference between the 2 paths is that one is allocating an extra temp (in the original code, both were doing that), and the ordering of the generated instructions (load-store-load-store vs. load-load-store-store), right?

Reply:
Looking at this again, I really don't see any reason to have 2 different code paths here.

Now that you've moved the ``set_cpu_vsrX`` calls out of the inner ``if`` clauses (which makes perfect sense), the only remaining difference between the two is the location of ``set_cpu_vsrh``. In the first code path, it's at the end of the block. In the second code path, it's at the middle of the block, between the two inner ``if`` clauses. If it's at the end (as in the first code path), you need the extra temporary - otherwise, a single temporary would be enough.

You can go with whichever is fastest:
- ``if`` at the end of the block + extra temporary; or
- ``if`` at the middle of the block without the extra temporary

But I still see no logical reason to keep both.

Reply:
Sorry. What I said didn't make sense. Probably still in post-vacation-brain-paralysis mode. :)

But here's another idea: looking at the second code path (the one without the ``unlikely``), which should be more efficient because it skips the 2nd temporary, it should only fail if the 2 following conditions are simultaneously true:
1. ``a->xt == a->xb``
2. ``(a->dm & 1) == 0``

And, in that case, it would be enough to just copy from ``a->xa`` and ``a->xb`` in the opposite order, while still requiring only one temporary:

```
t0 = tcg_temp_new_i64();

if (unlikely(a->xt == a->xb && (a->dm & 1) == 0)) {
    /* special case, read from the high part of VSR a->xb before overwriting it */

    get_cpu_vsrh(t0, a->xb);
    set_cpu_vsrl(a->xt, t0);

    if ((a->dm & 2) == 0) {
        get_cpu_vsrh(t0, a->xa);
    } else {
        get_cpu_vsrl(t0, a->xa);
    }
    set_cpu_vsrh(a->xt, t0);
} else {
    if ((a->dm & 2) == 0) {
        get_cpu_vsrh(t0, a->xa);
    } else {
        get_cpu_vsrl(t0, a->xa);
    }
    set_cpu_vsrh(a->xt, t0);

    if ((a->dm & 1) == 0) {
        get_cpu_vsrh(t0, a->xb);
    } else {
        get_cpu_vsrl(t0, a->xb);
    }
    set_cpu_vsrl(a->xt, t0);
}

tcg_temp_free_i64(t0);
```

Does that make sense?

Reply:
This approach (and others in this thread) seems to fail when xt == xb == xa, which is somewhat common because of the extended mnemonic `xxswapd`. I guess I'll keep the code as it is for now, and we can come back to these optimizations in the future if deemed necessary.